### PR TITLE
feat: Audit logs manager

### PR DESCRIPTION
**Pull request by jaw0r3k**
**Please describe the changes this PR makes and why it should be merged:**
1. Caches entries e.g. to check the admin who is banning a user *without unnecessary fetching*

**Status and versioning classification:**
- This PR changes the library's interface (methods or parameters added)
- I know how to update typings and have done so

### CODE REVIEW

File: packages/discord.js/src/managers/GuildAuditLogManager.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      }
+      const clone = existing._clone();
+      clone._patch(data);
+      return clone;
+    }
+
+    const entry = this.holds ? new this.holds(this.guild, data, ...extras) : data;
+    if (cache) this.cache.set(id ?? entry.id, entry);
+    return entry;
+  }
```
Comment (Syjalo) [inline packages/discord.js/src/managers/GuildAuditLogManager.js:45]: Why did you add this? `CachedManager` already has this method.

File: packages/discord.js/src/managers/GuildAuditLogManager.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+const CachedManager = require('./CachedManager');
+const { DiscordjsTypeError, ErrorCodes } = require('../errors');
+const GuildAuditLogs = require('../structures/GuildAuditLogs');
+const GuildAuditLogsEntry = require('../structures/GuildAuditLogsEntry');
+
+/**
+ * An extension for guild-specific application commands.
+ * @extends {CachedManager}
+ */
+class GuildAuditLogManager extends CachedManager {
```
Comment (Syjalo) [inline packages/discord.js/src/managers/GuildAuditLogManager.js:14]: Shouldn't the name be `GuildAuditLogEntryManager` since it holds `GuildAuditLogsEntry`?
Btw `GuildAuditLogsEntry` should be renamed to `GuildAuditLogEntry` in v15.

File: packages/discord.js/src/structures/Guild.js [hunk 3]
```diff
@@ -39,6 +38,12 @@ class Guild extends AnonymousGuild {
   constructor(client, data) {
     super(client, data, false);
 
+    /**
+     * A manager of the guild audit logs of this guild
+     * @type {GuildApplicationCommandManager}
+     */
+    this.auditLogs = new GuildAuditLogManager(this);
```
Comment (Syjalo) [inline packages/discord.js/src/structures/Guild.js:45]:
```suggestion
    this.auditLogEntries = new GuildAuditLogEntryManager(this);
```
Ditto

File: packages/discord.js/src/managers/GuildAuditLogManager.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    if (user) {
+      const userId = this.client.users.resolveId(user);
+      if (!userId) throw new DiscordjsTypeError(ErrorCodes.InvalidType, 'user', 'UserResolvable');
+      query.set('user_id', userId);
+    }
+
+    const data = await this.client.rest.get(Routes.guildAuditLog(this.id), { query });
+    return new GuildAuditLogs(this, data);
+  }
+}
```
Comment (Syjalo) [inline packages/discord.js/src/managers/GuildAuditLogManager.js:84]: I don't know if this should be here since this returns an `AuditLogs`, but the manager holds `AuditLogsEntry`.

**Review by Syjalo — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/Guild.js (hunk 3 above)
Comment (jaw0r3k) [reply packages/discord.js/src/structures/Guild.js:45]: not so sure about naming

File: packages/discord.js/src/managers/GuildAuditLogManager.js (hunk 4 above)
Comment (jaw0r3k) [reply packages/discord.js/src/managers/GuildAuditLogManager.js:84]: Like it takes some objects like commands 
which Entity uses
possibly they can be placed in a separate param
idk

Comment (jaw0r3k) [reply packages/discord.js/src/managers/GuildAuditLogManager.js:84]: maybe we can even delete GuildAuditLogs as it only has entires property
+ collections of objects

Comment (jaw0r3k) [reply packages/discord.js/src/managers/GuildAuditLogManager.js]: e90c94958761cbf7ba74139b912c5052b4951e48

File: packages/discord.js/src/managers/GuildAuditLogEntryManager.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+    if (user) {
+      const userId = this.client.users.resolveId(user);
+      if (!userId) throw new DiscordjsTypeError(ErrorCodes.InvalidType, 'user', 'UserResolvable');
+      query.set('user_id', userId);
+    }
+
+    const data = await this.client.rest.get(Routes.guildAuditLog(this.id), { query });
+
+    let notCachableObjects = {
```
Comment (jaw0r3k) [inline packages/discord.js/src/managers/GuildAuditLogEntryManager.js:86]: i am not too sure about it

File: packages/discord.js/src/managers/GuildAuditLogEntryManager.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    if (user) {
+      const userId = this.client.users.resolveId(user);
+      if (!userId) throw new DiscordjsTypeError(ErrorCodes.InvalidType, 'user', 'UserResolvable');
+      query.set('user_id', userId);
+    }
+
+    const data = await this.client.rest.get(Routes.guildAuditLog(this.id), { query });
+
+    const notCachableObjects = {
```
Comment (jaw0r3k) [inline packages/discord.js/src/managers/GuildAuditLogEntryManager.js:86]: i am not too sure about it

Comment (NotSugden) [reply packages/discord.js/src/managers/GuildAuditLogEntryManager.js:86]: nonCacheableObjects?
structures?
uncacheable?

**Review by kyranet — COMMENTED** [review]
I'll be honest here, but this is a no for me.

With TSR, I would personally like to move away from the excessive caching, as users can easily cache data on their way (plus, patching data with TSR would be even easier).

Besides, just like the message cache, the audit log one builds up, and can do so quite quickly, which increases memory usage. Users are aware of the settings to personalize the message cache, but they wouldn't be aware of this new manager, it can come up as unexpected and cause unexpected outages.

tl;dr: I don't see the benefit of this, you can cache this data easily by yourself.

**Comment by jaw0r3k** [comment]
> I'll be honest here, but this is a no for me.
> 
> With TSR, I would personally like to move away from the excessive caching, as users can easily cache data on their way (plus, patching data with TSR would be even easier).
> 
> Besides, just like the message cache, the audit log one builds up, and can do so quite quickly, which increases memory usage. Users are aware of the settings to personalize the message cache, but they wouldn't be aware of this new manager, it can come up as unexpected and cause unexpected outages.
> 
> tl;dr: I don't see the benefit of this, you can cache this data easily by yourself.

I agree
